### PR TITLE
Support search operators for commits search

### PR DESCRIPTION
**Pull request by typeless**
For now, support **auther:**, **committer:**, **after:** and **before:**,
When more than one prefix is supplied, the result is the union.
When different prefixes are supplied, the result is the intersection.

For example,
"author:alice author:bob"
=> the result is all commits authored by Alice OR Bob

"hello committer:alice after:2019-03-05 before:2019-03-10"
=> the result is all commits committed by Alice AND has the keyword during 2019-03-05~2019-03-10.
'hello' in the message.

Essentially, this is a wrapper for `git log` with the support of `--grep`, `--author`, `--committer`, `--after`, and `--before` options.

Note that there should NOT have any space after the colon(:) of the prefix.
For example,
"author:bill" => correct
"author: bill" => wrong

### CODE REVIEW

**Comment by lunny** [comment]
@typeless could you add some tests?

**Comment by typeless** [comment]
The test does not work yet.

**Comment by typeless** [comment]
Added some basic tests.

**Comment by typeless** [comment]
All tests passed. It is ready for review.

File: modules/git/commit.go [hunk 1]
```diff
@@ -215,8 +215,8 @@ func (c *Commit) CommitsBeforeUntil(commitID string) (*list.List, error) {
 }
 
 // SearchCommits returns the commits match the keyword before current revision
-func (c *Commit) SearchCommits(keyword string, all bool) (*list.List, error) {
-	return c.repo.searchCommits(c.ID, keyword, all)
+func (c *Commit) SearchCommits(keywords, authors, committers []string, after, before string, all bool) (*list.List, error) {
```
Comment (zeripath) [inline modules/git/commit.go]: At this point I'd probably say it's worth creating an options struct.

**Review by zeripath — APPROVED** [review]
Recommend making an options struct at the git module level. Otherwise lgtm

File: modules/git/commit.go (hunk 1 above)
Comment (typeless) [reply modules/git/commit.go]: Done.

**Review by zeripath — APPROVED** [review]

**Comment by typeless** [comment]
Hold on, I accidentally overwrite an old commit. :(
I am going to recover it.

**Comment by typeless** [comment]
Fixed.

File: modules/git/repo_commit.go [hunk 2]
```diff
@@ -237,30 +237,30 @@ func (repo *Repository) commitsByRange(id SHA1, page int) (*list.List, error) {
...
-	if len(committers) > 0 {
-		for _, v := range committers {
+	if len(opts.Committers) > 0 {
+		for _, v := range opts.Committers {
 			cmd.AddArguments("--committer=" + v)
 		}
 	}
-	if after != "" {
-		cmd.AddArguments("--after=" + after)
+	if opts.After != "" {
```
Comment (mrsdizzie) [inline modules/git/repo_commit.go]: Tiny style thing: This should be ```len(v) > 0``` to matched what is used elsewhere (and I believe preferred throughout the project). For both ```opts.After``` and ```opts.Before```

**Review by mrsdizzie — CHANGES_REQUESTED** [review]

File: modules/git/repo_commit.go (hunk 2 above)
Comment (typeless) [reply modules/git/repo_commit.go]: Done.

**Review by mrsdizzie — APPROVED** [review]

**Review by sapk — APPROVED** [review]

File: routers/repo/commit.go [hunk 3]
```diff
@@ -107,14 +107,42 @@ func SearchCommits(ctx *context.Context) {
...
 
-	keyword := strings.Trim(ctx.Query("q"), " ")
-	if len(keyword) == 0 {
+	query := strings.Trim(ctx.Query("q"), " ")
+	if len(query) == 0 {
 		ctx.Redirect(ctx.Repo.RepoLink + "/commits/" + ctx.Repo.BranchNameSubURL())
 		return
 	}
+
+	var keywords, authors, committers []string
```
Comment (lunny) [inline routers/repo/commit.go]: Could you create a new function named `NewSearchCommitsOptions(searchstring string) SearchCommitsOptions` to parse the content?

**Comment by lunny** [comment]
@typeless thanks for your PR. My comment is not a block that could be another PR.

**Comment by typeless** [comment]
@typeless okay, thanks.

File: routers/repo/commit.go (hunk 3 above)
Comment (typeless) [reply routers/repo/commit.go]: Done

**Review by lunny — APPROVED** [review]

**Pull request merged**